Src/Support/Google.Apis.Auth/OAuth2/ImpersonatedCredential.cs

```diff
@@ -0,0 +1,264 @@
+            var content = new StringContent(body, Encoding.UTF8, "application/json");
+            
+            var response = await HttpClient.PostAsync(TokenServerUrl, content, taskCancellationToken).ConfigureAwait(false);
+            var newToken = await TokenResponse.FromHttpResponseAsync(response, Clock, Logger).ConfigureAwait(false);
+            Token = newToken;
+            return true;
+        }
+
+        /// <inheritdoc/>
+        public Task<OidcToken> GetOidcTokenAsync(OidcTokenOptions options, CancellationToken cancellationToken = default)
+        {
+            options.ThrowIfNull(nameof(options));
+            // If at some point some properties are added to OidcToken that depend on the token having been fetched
+            // then initialize the token here.
+            TokenRefreshManager tokenRefreshManager = null;
+            tokenRefreshManager = new TokenRefreshManager(
+                ct => RefreshOidcTokenAsync(tokenRefreshManager, options, ct), Clock, Logger);
+            return Task.FromResult(new OidcToken(tokenRefreshManager));
+        }
+
+        private async Task<bool> RefreshOidcTokenAsync(TokenRefreshManager caller, OidcTokenOptions options, CancellationToken cancellationToken)
+        {
+            var bodyJson = new {
+                delegates = this.Delegates,
+                audience = options.TargetAudience,
+                includeEmail = true
+            };
+            var body = NewtonsoftJsonSerializer.Instance.Serialize(bodyJson);
+            var content = new StringContent(body, Encoding.UTF8, "application/json");
+            
+            var oidcTokenUrl = String.Format(GoogleAuthConsts.IamIdTokenEndpoint, TargetPrincipal);
+            var response = await HttpClient.PostAsync(oidcTokenUrl, content, cancellationToken).ConfigureAwait(false);
+            caller.Token = await TokenResponse.FromHttpResponseAsync(response, Clock, Logger).ConfigureAwait(false);
+            return true;
+        }
+
+        /// <summary>
+        /// Signs the provided bytes using the private key associated with the impersonated service account.
+        /// </summary>
+        /// <param name="bytesToSign">The bytes to sign.</param>
+        /// <param name="cancellationToken">Cancellation token to cancel operation.</param>
+        /// <returns>The signed bytes.</returns>
+        /// <exception cref="SignBlobResponseException">When signing request fails.</exception>
+        public async Task<byte[]> SignBytes(byte[] bytesToSign, CancellationToken cancellationToken = default) {
+            var bodyJson = new {
+                delegates = this.Delegates,
+                payload = Convert.ToBase64String(bytesToSign)
+            };
+            var body = new StringContent(NewtonsoftJsonSerializer.Instance.Serialize(bodyJson), Encoding.UTF8, "application/json");
+            
+            var signBlobUrl = String.Format(GoogleAuthConsts.IamSignEndpoint, TargetPrincipal);
```

Comment:
This makes me nervous - should we be doing any URL encoding of the target principal here?

Reply:
Encoding doesn't seem necessary, neither Python or Java does this, also the curl example in cloud doc uses it directly: https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials#sa-credentials-oauth

Target principal looks like `SA_NAME@PROJECT_ID.iam.gserviceaccount.com`, I think service account name and project id won't have weird characters.

Reply:
Well not in valid cases... but I don't think we're validating that the user has provided a reasonable email address.
It's *probably* okay... it just makes me a little uncomfortable.